src/phoenix/trace/filter.py

```diff
@@ -0,0 +1,180 @@
+        self._translated = _Translator(condition).visit(self._root)
+        ast.fix_missing_locations(self._translated)
+        self._compiled = compile(self._translated, filename="", mode="eval")
+
+    def __call__(self, span: Span) -> bool:
+        return cast(bool, eval(self._compiled, {}, {"span": span, "_MISSING": _Missing()}))
+
+
+def _replace_none_with_missing(
+    value: ast.expr,
+    as_str: bool = False,
+) -> ast.IfExp:
+    """E.g. `value` becomes
+    `_MISSING if (_MAYBE := value) is None else _MAYBE`
+    """
+    _store_MAYBE = ast.Name(id="_MAYBE", ctx=ast.Store())
+    _load_MAYBE = ast.Name(id="_MAYBE", ctx=ast.Load())
+    return ast.IfExp(
+        test=ast.Compare(
+            left=ast.NamedExpr(target=_store_MAYBE, value=value),
+            ops=[ast.Is()],
+            comparators=[ast.Constant(value=None)],
+        ),
+        body=ast.Name(id="_MISSING", ctx=ast.Load()),
+        orelse=_as_str(_load_MAYBE) if as_str else _load_MAYBE,
+    )
+
+
+def _as_str(value: ast.expr) -> ast.Call:
+    """E.g. `value` becomes `str(value)`"""
+    return ast.Call(func=ast.Name(id="str", ctx=ast.Load()), args=[value], keywords=[])
+
+
+def _ast_replacement(expression: str) -> ast.expr:
+    as_str = expression in (
+        "span.status_code",
+        "span.span_kind",
+        "span.parent_id",
+        "span.context.span_id",
+        "span.context.trace_id",
+    )
+    return _replace_none_with_missing(ast.parse(expression, mode="eval").body, as_str)
+
+
+def _allowed_replacements() -> Iterator[Tuple[str, ast.expr]]:
+    for source_segment, ast_replacement in {
+        "name": _ast_replacement("span.name"),
+        "status_code": _ast_replacement("span.status_code"),
+        "span_kind": _ast_replacement("span.span_kind"),
+        "parent_id": _ast_replacement("span.parent_id"),
+    }.items():
+        yield source_segment, ast_replacement
+        yield "span." + source_segment, ast_replacement
+    for source_segment, ast_replacement in {
+        "span_id": _ast_replacement("span.context.span_id"),
+        "trace_id": _ast_replacement("span.context.trace_id"),
+    }.items():
+        yield source_segment, ast_replacement
+        yield "context." + source_segment, ast_replacement
+        yield "span.context." + source_segment, ast_replacement
+    for source_segment, ast_replacement in {
+        field_name: _ast_replacement(f"span.attributes.get('{field_name}')")
+        for field_name in (
+            getattr(semantic_conventions, variable_name)
+            for variable_name in dir(semantic_conventions)
+            if variable_name.isupper()
+            and variable_name.startswith(("RETRIEVAL", "EMBEDDING", "LLM", "TOOL"))
+        )
+    }.items():
+        yield source_segment, ast_replacement
+        yield "attributes." + source_segment, ast_replacement
+        yield "span.attributes." + source_segment, ast_replacement
+
+
+class _Translator(ast.NodeTransformer):
```

Comment:
Keeping this private via _ just makes the code just harder to read - I'm not sure it's worth it